pkg/controller/projectreference/projectreference_adapter.go

```diff
@@ -574,6 +576,63 @@ func (r *ReferenceAdapter) SetIAMPolicy(serviceAccountEmail string) error {
 
 }
 
+// SetProjectReferenceCondition sets a condition on a ProjectReference's status
+func (r *ReferenceAdapter) SetProjectReferenceCondition(status corev1.ConditionStatus, reason string, message string) error {
+	conditions := &r.projectReference.Status.Conditions
+	conditionType := gcpv1alpha1.ProjectReferenceConditionError
+	now := metav1.Now()
+	existingConditions := r.findProjectReferenceCondition()
+	if existingConditions == nil {
+		if status == corev1.ConditionTrue {
+			*conditions = append(
+				*conditions,
+				gcpv1alpha1.ProjectReferenceCondition{
+					Type:               conditionType,
+					Status:             status,
+					Reason:             reason,
+					Message:            message,
+					LastTransitionTime: now,
+					LastProbeTime:      now,
+				},
+			)
+		}
+	} else {
+		if existingConditions.Status != status {
+			existingConditions.LastTransitionTime = now
+		}
+		existingConditions.Status = status
+		existingConditions.Reason = reason
+		existingConditions.Message = message
+		existingConditions.LastProbeTime = now
+	}
+
+	return r.statusUpdate()
+}
```

Comment:
I like that we're doing this similar to the ProjectReference handling, however, this is a siginificant amount of code duplication (https://github.com/openshift/gcp-project-operator/blob/6e30cdb0d2cf9e9fa03f40e2a09858f6a47f2c5f/pkg/controller/projectclaim/projectclaimadapter.go#L201), I think it would be nice to have this unified, if possible. Otherwise we risk the two functions evolve independently, which might become a hotbed for bugs..

Reply:
@NautiluX that function's logic is exported to `util.go` and callable from either projectreference or projectclaim. Can you have a look at that again?